Functions/Quantification/xASL_quant_ASL.m

```diff
@@ -153,29 +159,53 @@
         error('Wrong PLD definition!');
     end
 
+
+    %% 2. Run BASIL quantification
     if bUseBasilQuantification
         % Here we perform FSL quantification
-		
-		[PWI, ATT, ABV, Tex] = xASL_quant_FSL(PWI, x);
+
+        % #1543 We may want to move the NaN extrapolation here
+
+		[PWI, ATT, ABV, Tex] = xASL_quant_FSL(x.P.Path_PWI4D, x); % #1543 Quick and dirty fix for the NaN extrapolation at the start of xASL_quant_FSL
 		
 		% If resultFSL is not 0, something went wrong
         % This will issue a warning inside xASL_quant_FSL
 	else
         % This part should only run when we don't use FSL BASIL/FABBER
         % or as a fallback when FSL BASIL/FABBER crashed
         
-        %% 2    Label decay scale factor for single (blood T1) - or dual-compartment (blood+tissue T1) model, CASL or PASL
-        if isfield(x.Q,'LabelingType') && isfield(x.Q,'LabelingDuration')
+        % First, we average PWI4D into PWI, for single-PLD only
+        % (later, when we have multi-PLD, multi-echo, or multi-labeling quantification here as well,
+        % then we could use PWI3D here)
+
+        if isfield(x.Q, 'SaveCBF4D') && x.Q.SaveCBF4D
+            PWI = PWI4D;
+            fprintf('%s\n', 'Quantifying CBF in 4D');
+            % In this case, we want to save a quantified version of PWI4D
+        else
+            PWI = xASL_im_ASLSubtractionAveraging(x, [], PWI4D);
+        end
+
+        % #997 Not sure if this is the best location for this checking
+        if ~x.modules.asl.bMultiTE && length(x.Q.uniqueEchoTimes)>1
```

Comment:
I would remove this whole paragraph. Because at the start of the ASL module, bmultiTE is set to TRUE only when we have multiple echos. bMultiTE is our internal parameter, it is not a parameter from dataPar.json. So this case will never happen. It's probably an old code that was written there before we had this proper check in ASL module.

Reply:
Weird, I try to find this but cannot find it, did you already remove this?

Without being able to inspect the code: I agree with you except that bMultiTE is decided in the pipeline where uniqueEchoTime is loaded from JSON, and potentially changed by the user if they want to manually adapt PWI4D for quantification (e.g., manual merging)

Reply:
I kept the code it - I only made this comment because I was looking for all #997 comments. It is not even outdated, so it is still on the same line. 

Currently, the user cannot change bMultiTE himself. If you want the user to have this ability. then you have to describe how this would work and at what stage. And we can implement that. But we can't expect the user to randomly touch the data, ExploreASL is an automated pipeline and if users keep interfering with data in an unpredictable way, then the entire pipeline will crash. I know you advocate for modularity, but the entire pipeline has certain dependencies and the user cannot randomly tamper with derivatives.

So currently, this paragraph is useless. Unless you want to propose major changes on how is this handled.

Reply:
Fixed in 3b440c54

Reply:
With that I agree. In that case, we need to add bMultiTE as a possible parameter to dataPar.json, add it to the documentation and explain that this is used to turn things off. And we have to make sure that it can correctly propagate until here and that things run really like that...

And then I'll add that warning again and say - "warning, not implemented yet".

Is that OK? My point was - it wasn't illegal, but it can't possibly happen :) Now I will implement the possibility to turn it OFF. And I'll make a new issue that properly checks that turning off really does the correct job for DEBBIE - that things are really done single-TE without breaking? Sounds good?

Reply:
Yes, indeed.

But then we turn it around right, for DEBBIE you want to turn bMultiTEQuantification (that is probably what we should call it? Because this boolean doesn't change the rest of the ASL module?) on?

And you are right, by default we want bMultiTEQuantification detected for multi-TE sequences and running this. So essentially you use it to turn it off :)

Reply:
OK. will do that, change the name, add to documentation.

Reply:
Option `bMultiTEQuantification` created that allows to turn multiTE quantification. 

Added to the documentaiton.

Reintroduced the warning.

See 04984941

Reply:
Why should we remove bMultiTE - that's an indicator that we have multiTE or not. You have very specifically asked for this parameter 1 year ago.

We don't need bMultiPLDQuantification, because we're not going to turn multi-PLD quantification ON/OFF. It makes no sense to do a single-PLD quantification of a multi-PLD sequence. And anyway, you would have to select which PLDs. So this makes no sense.

Reply:
There can be other reasons like a dual-echo ASL sequence, to only quantify the first echo.

That you have select which PLD is always the case, if you have single or multi-PLDs. We currently circumvent that by Amnah's modeling, but that is not a general motivation. 

But see my other comment. I agree with you that multi-TE quantification should be defaulted to. But there should be the option to disable this when you have multiple echoes because of a fMRI or DTI combination with ASL and you only want to process the first (like dual-echo ASL).